src/lab16.py

```diff
@@ -278,69 +308,49 @@ def __init__(self, node, parent, previous, frame):
         self.y = None
         self.width = None
         self.height = None
+        self.zoom = None
 
         self.dirty_children = True
-        self.dirty_zoom = True
-        self.zoom = None
+        self.zoom_field = DependentField(self, "zoom")
         self.dirty_style = True
-        self.dirty_width = True
-        self.dirty_height = True
-        self.dirty_x = True
-        self.dirty_y = True
+        self.width_field = DependentField(self, "width")
+        self.x_field = DependentField(self, "x")
+        self.y_field = DependentField(self, "y")
+        self.height_field = DependentField(self, "height")
         self.dirty_descendants = True
 
     def layout(self):
-        if self.dirty_zoom:
-            self.zoom = self.parent.zoom
-            for child in self.children:
-                mark_dirty(child)
-                child.dirty_zoom = True
-            mark_dirty(self)
-            self.dirty_width = True
-            self.dirty_zoom = False
+        parent_zoom = self.zoom_field.read(self.parent.zoom_field)
+        self.zoom = self.zoom_field.set(parent_zoom)
 
         if self.dirty_style:
             self.dirty_children = True
-            self.dirty_width = True
+            self.width_field.dirty = True
```

Comment:
This is the only place you explicitly set dirty to true. Is there something special about it?